pkg/controllers/cspi-controller/handler.go

```diff
@@ -0,0 +1,300 @@
+		err = oc.Create(cspi)
+		if err != nil {
+			// We will try to create it in next event
+			c.recorder.Event(cspi,
+				corev1.EventTypeWarning,
+				string(common.FailureCreate),
+				fmt.Sprintf("Failed to create pool due to '%s'", err.Error()))
+
+			_ = oc.Delete(cspi)
+			common.SyncResources.Mux.Unlock()
+			return nil
+		}
+		common.SyncResources.Mux.Unlock()
+
+		c.recorder.Event(cspi,
+			corev1.EventTypeNormal,
+			string(common.SuccessCreated),
+			fmt.Sprintf("Pool created successfully"))
+
+		err = c.update(cspi)
+		if err != nil {
+			c.recorder.Event(cspi,
+				corev1.EventTypeWarning,
+				string(common.FailedSynced),
+				err.Error())
+		}
+		return nil
+
+	}
+	common.SyncResources.Mux.Unlock()
+	return nil
+}
+
+func (c *CStorPoolInstanceController) destroy(cspi *cstor.CStorPoolInstance) error {
+	var phase cstor.CStorPoolInstancePhase
+
+	if !util.ContainsString(cspi.Finalizers, apiscspc.PoolProtectionFinalizer) {
+		return nil
+	}
+	// Instantiate the pool operation config
+	oc := zpool.NewOperationsConfig().WithKubeClientSet(c.kubeclientset).WithOpenEBSClient(c.clientset)
+	// DeletePool is to delete cstor zpool.
+	// It will also clear the label for relevant disk
+	err := oc.Delete(cspi)
+	if err != nil {
+		c.recorder.Event(cspi,
+			corev1.EventTypeWarning,
+			string(common.FailureDestroy),
+			fmt.Sprintf("Failed to delete pool due to '%s'", err.Error()))
+		phase = cstor.CStorPoolStatusDeletionFailed
+		goto updatestatus
+	}
+
+	// removeFinalizer is to remove finalizer of cStorPoolInstance resource.
+	err = c.removeFinalizer(cspi)
+	if err != nil {
+		// Object will exist. Let's set status as offline
+		klog.Errorf("removeFinalizer failed %s", err.Error())
+		phase = cstor.CStorPoolStatusDeletionFailed
+		goto updatestatus
+	}
+	klog.Infof("Pool %s deleted successfully", cspi.Name)
+	return nil
+
+updatestatus:
+	cspi.Status.Phase = phase
+	if _, er := zpool.OpenEBSClient.
+		CstorV1().
+		CStorPoolInstances(cspi.Namespace).
+		Update(cspi); er != nil {
+		klog.Errorf("Update failed %s", er.Error())
+	}
+	return err
+}
+
+func (c *CStorPoolInstanceController) update(cspi *cstor.CStorPoolInstance) error {
+	oc := zpool.NewOperationsConfig().WithKubeClientSet(c.kubeclientset).WithOpenEBSClient(c.clientset)
+	cspi, err := oc.Update(cspi)
+	if err != nil {
+		return errors.Errorf("Failed to update pool due to %s", err.Error())
+	}
+	return c.updateStatus(cspi)
+}
+
+func (c *CStorPoolInstanceController) updateStatus(cspi *cstor.CStorPoolInstance) error {
+	// ToDo: Use the status from the cspi object that is passed in arg else other fields
+	// might get lost.
+	var status cstor.CStorPoolInstanceStatus
+	var err error
+	pool := zpool.PoolName(cspi)
+
+	state, er := zpool.GetPropertyValue(pool, "health")
+	if er != nil {
+		err = zpool.ErrorWrapf(err, "Failed to fetch health")
+	} else {
+		status.Phase = cstor.CStorPoolInstancePhase(state)
+	}
+
+	freeSize, er := zpool.GetPoolCapacity(pool, "free")
```

Comment:
IMO: Instead of making multiple calls to zfs good to get all the details in one call.